# Enrichr Results Compilation

Importing necessary packages, loading library segments, and setting up data for the benchmark dataset.

In [2]:
### Sourcing tools from GEO ###
    
source("http://www.bioconductor.org/biocLite.R")
    
BiocManager::install("GenomicRanges")
BiocManager::install("rtracklayer")
BiocManager::install("usethis")
library(GenomicRanges)
library(rtracklayer)
library(usethis)
    
### Installing package dependencies ###
    
BiocManager::install("devtools")
BiocManager::install("roxygen2")
library(devtools)
library(roxygen2)

Warning message:
“'BiocInstaller' and 'biocLite()' are deprecated, use the 'BiocManager' CRAN
  package instead.”Bioconductor version 3.8 (BiocInstaller 1.32.1), ?biocLite for help
Warning message in file(con, "r"):
“URL 'https://bioconductor.org/config.yaml': status was 'Timeout was reached'”Bioconductor version 3.8 (BiocManager 1.30.4), R 3.5.2 (2018-12-20)
Installing package(s) 'GenomicRanges'



The downloaded binary packages are in
	/var/folders/hm/c3_fjypn62v5xh5b5ygv267m0000gn/T//Rtmp5xnwAB/downloaded_packages


Update old packages: 'callr', 'checkmate', 'deSolve', 'digest', 'dplyr',
  'ellipsis', 'emmeans', 'gam', 'git2r', 'haven', 'lavaan', 'mclust', 'pillar',
  'plotmo', 'processx', 'quantreg', 'recipes', 'remotes', 'rlang', 'rngtools',
  'robust', 'sf', 'tinytex', 'vdiffr', 'xfun'
Bioconductor version 3.8 (BiocManager 1.30.4), R 3.5.2 (2018-12-20)
Installing package(s) 'rtracklayer'



The downloaded binary packages are in
	/var/folders/hm/c3_fjypn62v5xh5b5ygv267m0000gn/T//Rtmp5xnwAB/downloaded_packages


Update old packages: 'callr', 'checkmate', 'deSolve', 'digest', 'dplyr',
  'ellipsis', 'emmeans', 'gam', 'git2r', 'haven', 'lavaan', 'mclust', 'pillar',
  'plotmo', 'processx', 'quantreg', 'recipes', 'remotes', 'rlang', 'rngtools',
  'robust', 'sf', 'tinytex', 'vdiffr', 'xfun'
Bioconductor version 3.8 (BiocManager 1.30.4), R 3.5.2 (2018-12-20)
Installing package(s) 'usethis'



  There is a binary version available but the source version is later:
        binary source needs_compilation
usethis  1.5.0  1.5.1             FALSE



installing the source package ‘usethis’

Update old packages: 'callr', 'checkmate', 'deSolve', 'digest', 'dplyr',
  'ellipsis', 'emmeans', 'gam', 'git2r', 'haven', 'lavaan', 'mclust', 'pillar',
  'plotmo', 'processx', 'quantreg', 'recipes', 'remotes', 'rlang', 'rngtools',
  'robust', 'sf', 'tinytex', 'vdiffr', 'xfun'
Loading required package: stats4
Loading required package: BiocGenerics
Loading required package: parallel

Attaching package: ‘BiocGenerics’

The following objects are masked from ‘package:parallel’:

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB

The following objects are masked from ‘package:stats’:

    IQR, mad, sd, var, xtabs

The following objects are masked from ‘package:base’:

    Filter, Find, Map, Position, Reduce, anyDuplicated, append,
    as.data.frame, basename, cbind, colMeans, colSums, colnames,
    dirname, do.call, duplicated,


The downloaded binary packages are in
	/var/folders/hm/c3_fjypn62v5xh5b5ygv267m0000gn/T//Rtmp5xnwAB/downloaded_packages


Update old packages: 'callr', 'checkmate', 'deSolve', 'digest', 'dplyr',
  'ellipsis', 'emmeans', 'gam', 'git2r', 'haven', 'lavaan', 'mclust', 'pillar',
  'plotmo', 'processx', 'quantreg', 'recipes', 'remotes', 'rlang', 'rngtools',
  'robust', 'sf', 'tinytex', 'vdiffr', 'xfun'
Bioconductor version 3.8 (BiocManager 1.30.4), R 3.5.2 (2018-12-20)
Installing package(s) 'roxygen2'



The downloaded binary packages are in
	/var/folders/hm/c3_fjypn62v5xh5b5ygv267m0000gn/T//Rtmp5xnwAB/downloaded_packages


Update old packages: 'callr', 'checkmate', 'deSolve', 'digest', 'dplyr',
  'ellipsis', 'emmeans', 'gam', 'git2r', 'haven', 'lavaan', 'mclust', 'pillar',
  'plotmo', 'processx', 'quantreg', 'recipes', 'remotes', 'rlang', 'rngtools',
  'robust', 'sf', 'tinytex', 'vdiffr', 'xfun'


In [8]:
ChIPSeqDataMaster <- read.table(file = "/Users/mei/Desktop/benchmarks/genomic_range/data/GSA_ChIP_Seq_Master_Table.txt", sep = '\t', header = TRUE, quote = "")

In [10]:
## The following code creates a list of samples for which we need to extract the BED files for.
    
ChIPSeqSamples<-as.character(ChIPSeqDataMaster$GSM)
    
## Initializing list for storing BED files and the consecutive GRanges objects.
Samples_in_BED = list()
    
for(i in 1:length(ChIPSeqSamples))
    {
      Samples_in_BED[[i]] <- read.table(paste0("/Users/mei/LabWork/GSAChIPSeqBenchmarkDataBase/regen/",paste0(eval(parse(text="ChIPSeqSamples[i]")),".bed")), sep = "\t", header = FALSE)
      Samples_in_BED[[i]] <- Samples_in_BED[[i]][,1:3]
      colnames(Samples_in_BED[[i]]) <- c("chrom", "start", "end")
      Samples_in_BED[[i]] <- Samples_in_BED[[i]][order(Samples_in_BED[[i]]$chrom),]
      Samples_in_BED[[i]] <- GRanges(Samples_in_BED[[i]]$chrom, IRanges(Samples_in_BED[[i]]$`start`, Samples_in_BED[[i]]$`end`))
      genome(Samples_in_BED[[i]]) <- "hg19"
    }
    
## Saving BED files as GRanges objects ##
names(Samples_in_BED) <- ChIPSeqSamples

## GO Results

The results from Enrichr were manually curated for the samples from the benchmark dataset and stored at distinct locations for KEGG and GO enrichments; the folders were named **KEGG_2016** and **GO_BP_2016**, respectively. The code has been sourced from **Enrichr_Results_Compilaton.R** and **Enrichr_Preprocessing.R** functions available at https://github.com/mora-lab/benchmarks/tree/master/genomic_range/R.

In [12]:
tryCatch( {enrichr_go_samples <- read.table("/Users/mei/LabWork/GSAChIPSeqBenchmarkDataBase/Results/enrichr/GO_BP_2018/file_names.txt")}
          ,error = function(e){ print("File not found"); break;}
          ,finally = function (f){next;})

In [13]:
class(enrichr_go_samples)

[1] "data.frame"

In [14]:
head(enrichr_go_samples)

V1            
1 GSE84618.txt  
2 GSM1847178.txt
3 GSM2058015.txt
4 GSM2058016.txt
5 GSM2058017.txt
6 GSM2058018.txt

As you can see here, **enrichr_go_samples** is a dataframe with one column holding the samples whose corresponding results are placed in the folder. We want to extract these results. For doing so, we shall begin with pruning the file names by removing extension- *.txt*.

In [15]:
enrichr_go_samples <- as.character(enrichr_go_samples$V1)
for(i in 1:length(enrichr_go_samples)){enrichr_go_samples[i] <- substr(enrichr_go_samples[i],1,nchar(enrichr_go_samples[i])-4)}

In [16]:
head(enrichr_go_samples)

[1] "GSE84618"   "GSM1847178" "GSM2058015" "GSM2058016" "GSM2058017"
[6] "GSM2058018"

Now that we have the precise sample labels, let us define a list **enrichr_go** that shall hold the enrichr results (dataframe), for each sample. 

In [19]:
enrichr_go <- list()

for (i in 1: length(ChIPSeqSamples))
{
  for(j in 1:length(enrichr_go_samples))
  {
    if(enrichr_go_samples[j] == ChIPSeqSamples[i])
    {
      enrichr_go[[j]] <-read.table(paste0("/Users/mei/LabWork/GSAChIPSeqBenchmarkDataBase/Results/enrichr/GO_BP_2018/",paste0(eval(parse(text='ChIPSeqSamples[i]')),".txt")), sep = '\t', header = TRUE, quote = "", fill = TRUE)
    }
  }
}

In [20]:
class(enrichr_go)

[1] "list"

So, eventually **enrichr_go** is a list of dataframes.

In [21]:
head(enrichr_go[[1]])

Term                                                                  
1 regulation of skeletal muscle contraction (GO:0014819)                
2 cell-cell junction maintenance (GO:0045217)                           
3 regulation of transforming growth factor beta2 production (GO:0032909)
4 regulation of histone H3-K9 acetylation (GO:2000615)                  
5 regulation of Golgi organization (GO:1903358)                         
6 chondroitin sulfate catabolic process (GO:0030207)                    
  Overlap P.value      Adjusted.P.value Old.P.value  Old.Adjusted.P.value
1 3/7     0.0037479580 0.9441197        0.0052339067 0.7605585           
2 4/11    0.0015450556 0.9441197        0.0019520713 0.7605585           
3 2/7     0.0443520949 0.9441197        0.0427839212 0.7605585           
4 3/8     0.0057748139 0.9441197        0.0069970387 0.7605585           
5 5/15    0.0006100561 0.9441197        0.0007217959 0.7605585           
6 4/14    0.0041556430 0.9441197        0.0040004439 0.7605585           
  Z.score   Combined.Score Genes                          
1 -3.158475 17.64496       CAV3;DMPK;KCNJ2                
2 -2.577299 16.68207       CAMSAP3;NLGN2;PLEKHA7;CD177    
3 -4.395784 13.69548       SMAD4;CDH3                     
4 -2.630837 13.55999       SMAD4;CHEK1;BRCA1              
5 -1.822244 13.48818       CAMSAP3;MAP2K1;STX18;RBSN;MAPK3
6 -2.408644 13.20729       BCAN;HYAL1;NCAN;ARSB

We observe an element from the list, and see that it outputs quite a bit og information. Following the scope of our analysis, we shall only require the **Term** and **P.value** parameters. So, the remaining columns are filtered out.

In [22]:
enrichr_go_results <- list()
for (i in 1:length(enrichr_go))
{
  enrichr_go_results[[i]] <- enrichr_go[[i]][,c(1,3)]
}

names(enrichr_go_results) <- as.character(enrichr_go_samples)

## KEGG Results

We shall follow the exact same protocol for KEGG results as well.

In [25]:
## Same protocol for ENRICHR KEGG results too. 


enrichr_kegg <- list()

enrichr_kegg_samples <- read.table("/Users/mei/LabWork/GSAChIPSeqBenchmarkDataBase/Results/enrichr/KEGG_2016/file_names.txt")
enrichr_kegg_samples <- as.character(enrichr_kegg_samples$V1)
for(i in 1:length(enrichr_kegg_samples)){enrichr_kegg_samples[i] <- substr(enrichr_kegg_samples[i],1,nchar(enrichr_kegg_samples[i])-4)}



enrichr_kegg <- list()

for (i in 1: length(ChIPSeqSamples))
{
  for(j in 1:length(enrichr_kegg_samples))
  {
    if(enrichr_kegg_samples[j] == ChIPSeqSamples[i])
    {
      enrichr_kegg[[j]] <-read.table(paste0("/Users/mei/LabWork/GSAChIPSeqBenchmarkDataBase/Results/enrichr/KEGG_2016/",paste0(eval(parse(text='ChIPSeqSamples[i]')),".txt")), sep = '\t', header = TRUE, quote = "", fill = TRUE)
    }
  }
}


## Condensed Results

enrichr_kegg_results <- list()
for (i in 1:length(enrichr_kegg))
{
  enrichr_kegg_results[[i]] <- enrichr_kegg[[i]][,c(1,3)]
}

names(enrichr_kegg_results) <- as.character(enrichr_kegg_samples)


Now that we have the individual enrichment results from GO and KEGG ontologies, we shall proceed with combining the results for each sample to engender a consolidated set. Prior to that, let's check if the sample count is consistent for both cohorts.

In [28]:
# for consistency with Chipenrich, Broadenrich, Seq2pathway results' variable nomenclature.
enrichr_go_results_shredded <- enrichr_go_results
enrichr_kegg_results_shredded <- enrichr_kegg_results

count <- 0
for (i in 1:length(names(enrichr_go_results_shredded)))
{
  for(j in 1:length(names(enrichr_kegg_results_shredded)))
  {
    if(names(enrichr_go_results_shredded)[i] == names(enrichr_kegg_results_shredded)[j])
    {
      count <- count+1
    }
  }
}
print (count)

[1] 106


Cool ! That shows that we have results from  both KEGG and GO for all samples; number of samples in our benchmark dataset is 106.

In [32]:
## Refining Enrichr KEGG and GO results | Extracting "hsa*****" and "GO:*****" terms.
#install.packages("stringr", repos = "https://mirrors.tuna.tsinghua.edu.cn/CRAN/")
#BiocManager::install("stringr")
#library(stringr)
enrichr_go_terms_extracted_results_shredded <- list()
enrichr_go_terms_extracted_results_shredded <- enrichr_go_results_shredded
for (i in 1:length(enrichr_go_results_shredded))
{
  enrichr_go_terms_extracted_results_shredded[[i]]$Term <- str_extract(string = print(eval(parse(text=paste0("enrichr_go_results_shredded$",paste0(eval(parse(text="names(enrichr_go_results_shredded)[i]")),"$Term"))))), pattern = "GO:[0-9]+")
}

enrichr_kegg_terms_extracted_results_shredded <- list()
enrichr_kegg_terms_extracted_results_shredded <- enrichr_kegg_results_shredded
for (i in 1:length(enrichr_kegg_results_shredded))
{
  enrichr_kegg_terms_extracted_results_shredded[[i]]$Term <- str_extract(string = print(eval(parse(text=paste0("enrichr_kegg_results_shredded$",paste0(eval(parse(text="names(enrichr_kegg_results_shredded)[i]")),"$Term"))))), pattern = "hsa[0-9]+")
}


Warning message in get(var, frame, inherits = FALSE):
“restarting interrupted promise evaluation”Warning message in get(var, frame, inherits = FALSE):
“internal error -3 in R_decompress1”Warning message:
“restarting interrupted promise evaluation”Warning message:
“internal error -3 in R_decompress1”

ERROR: Error: lazy-load database '/Library/Frameworks/R.framework/Versions/3.5/Resources/library/stringr/R/stringr.rdb' is corrupt
